public-dashboard-client/src/viz-recidivism-rates/VizRecidivismRates.js

```diff
@@ -37,30 +46,70 @@ function typeCast(recidivismRecord) {
   };
 }
 
-function prepareChartData({ data, selectedCohorts }) {
-  return Array.from(
-    group(data.map(typeCast), (d) => d.releaseCohort),
-    ([key, value]) => {
-      return {
-        label: key,
-        coordinates: value,
-      };
-    }
-  )
-    .map(assignOrderedDatavizColor)
-    .filter((record) => {
-      if (!selectedCohorts) {
-        return true;
+/**
+ * When multiple cohorts are selected, or a single cohort and the `total` dimension,
+ * will return one data series per cohort.
+ * Otherwise (i.e., a single cohort and some dimensional breakdown is selected),
+ * will return one data series per demographic subgroup.
+ */
+function prepareChartData({ data, dimension, selectedCohorts }) {
```

Comment:
This might be a nice place to add a test. There are a few branches of logic so a couple of tests to show the chart data at the end of each branch might be helpful.

Reply:
I did add some tests here but I have mixed feelings about them. Exporting and testing this function itself felt like too much of a low-level implementation detail, since it's so specific to the Semiotic API. Testing the rendered output is also not super obvious since it's mainly just, like, drawing a bunch of lines? I settled on testing the label text that Semiotic renders, since it should reflect the data input in the same way that the chart marks do, but it's still pretty idiosyncratic ... e.g. if we swapped in a different library that produced an equivalent-looking chart I have low confidence that these tests would still pass, since chart labeling is still not a well established web standard.

Also I had to do something kind of funky to get the data fixture into the test environment? I actually think it's kind of a fun solution honestly but it is definitely quick-and-dirty.